### DISCUSSION TITLE
`NetworkBehaviour` derive macro now requires `From<Void>` for corresponding `OutEvent`

### DISCUSSION THREAD

**Discussion by drHuangMHT** [libp2p/rust-libp2p#3784] [Q&A] [answered]
Why `From<Void>`? `Void` means something never happens. Is it intentional? Errors start popping up when using latest `libp2p` in this repo.

**Comment by thomaseizinger** [comment, marked as answer]
It depends which `NetworkBehaviour`s you are composing. Some of them never emit an event and thus have their `OutEvent` set to `Void`.

**Reply by drHuangMHT** [reply]
Found using `keep_alive` behaviour whose `OutEvent` is `Void`. But this problem did not occur before so I'm not sure what happened.

**Reply by thomaseizinger** [reply]
What do you mean with "before"? Before you upgraded? Or started using the keep alive behaviour?

**Reply by drHuangMHT** [reply]
Before I upgraded. It's not an issue in `0.50` or `0.51` when I started using `libp2p`. The code compiles just fine with `keep_alive` without deriving `From<Void>`. Confirmed with code on `0.50` with feature `full`.

**Reply by thomaseizinger** [reply]
To answer your initial question, yes it is intentional. The custom-derive doesn't know the semantics of the code so it cannot "skip" this implementation despite it being not possible at runtime. The implementation is very simple though:

```rust
impl From<Void> for Event {
	fn(v: Void) -> Event {
		void::unreachable(v)
	}
}
```

**Reply by drHuangMHT** [reply]
But `libp2p` does not re-export `Void`, is that also intentional?

**Reply by thomaseizinger** [reply]
`void` is already 1.0.0 so we don't export it as we consider it stable enough to just depend on it.

**Reply by thomaseizinger** [reply]
Also see https://github.com/libp2p/rust-libp2p/discussions/3977.
